br/pkg/lightning/config/config.go

```diff
@@ -154,22 +154,113 @@ type DBStore struct {
 	UUID         string         `toml:"-" json:"-"`
 }
 
+// adjust assigns default values and check illegal values. The arguments must be
+// adjusted before calling this function.
+func (d *DBStore) adjust(
+	ctx context.Context,
+	i *TikvImporter,
+	s *Security,
+	tlsObj *common.TLS,
+) error {
+	if i.Backend == BackendLocal {
+		if d.BuildStatsConcurrency == 0 {
+			d.BuildStatsConcurrency = defaultBuildStatsConcurrency
+		}
+		if d.IndexSerialScanConcurrency == 0 {
+			d.IndexSerialScanConcurrency = defaultIndexSerialScanConcurrency
+		}
+		if d.ChecksumTableConcurrency == 0 {
+			d.ChecksumTableConcurrency = defaultChecksumTableConcurrency
+		}
+	}
+	var err error
+	d.SQLMode, err = mysql.GetSQLMode(d.StrSQLMode)
+	if err != nil {
+		return common.ErrInvalidConfig.Wrap(err).GenWithStack("`mydumper.tidb.sql_mode` must be a valid SQL_MODE")
+	}
+
+	if d.Security == nil {
+		d.Security = s
+	}
+
+	switch d.TLS {
+	case "skip-verify", "preferred":
+		if d.Security.TLSConfig == nil {
+			/* #nosec G402 */
+			d.Security.TLSConfig = &tls.Config{
+				MinVersion:         tls.VersionTLS10,
+				InsecureSkipVerify: true,
+				NextProtos:         []string{"h2", "http/1.1"}, // specify `h2` to let Go use HTTP/2.
+			}
+			d.Security.AllowFallbackToPlaintext = true
+		}
+	case "cluster":
+		if len(s.CAPath) == 0 {
+			return common.ErrInvalidConfig.GenWithStack("cannot set `tidb.tls` to 'cluster' without a [security] section")
+		}
+	case "", "false":
+		d.TLS = "false"
+	default:
+		return common.ErrInvalidConfig.GenWithStack("unsupported `tidb.tls` config %s", d.TLS)
+	}
+
+	mustHaveInternalConnections := i.Backend == BackendLocal
+	// automatically determine the TiDB port & PD address from TiDB settings
+	if mustHaveInternalConnections && (d.Port <= 0 || len(d.PdAddr) == 0) {
+		var settings tidbcfg.Config
+		err = tlsObj.GetJSON(ctx, "/settings", &settings)
+		if err != nil {
+			return common.ErrInvalidConfig.Wrap(err).GenWithStack("cannot fetch settings from TiDB, please manually fill in `tidb.port` and `tidb.pd-addr`")
+		}
+		if d.Port <= 0 {
+			d.Port = int(settings.Port)
+		}
+		if len(d.PdAddr) == 0 {
+			pdAddrs := strings.Split(settings.Path, ",")
+			d.PdAddr = pdAddrs[0] // FIXME support multiple PDs once importer can.
+		}
+	}
+
+	if d.Port <= 0 {
+		return common.ErrInvalidConfig.GenWithStack("invalid `tidb.port` setting")
+	}
+
+	if mustHaveInternalConnections && len(d.PdAddr) == 0 {
+		return common.ErrInvalidConfig.GenWithStack("invalid `tidb.pd-addr` setting")
+	}
+	return nil
+}
+
+// Routes is a alias of []*router.TableRule. It's used to attach method to []*router.TableRule.
+type Routes []*router.TableRule
+
+func (r *Routes) adjust(m *MydumperRuntime) error {
+	for _, rule := range *r {
+		if !m.CaseSensitive {
+			rule.ToLower()
+		}
+		if err := rule.Valid(); err != nil {
+			return common.ErrInvalidConfig.Wrap(err).GenWithStack("file route rule is invalid")
```

Comment:
table route?

Reply:
In lightning it's indeed called file route. Maybe the data source of lighting are files.